### ISSUE TITLE
TCP backend: use laik_log instead of own log/error stuff

### ISSUE THREAD

**Issue by weidendo** [envelope-project/laik#125]

**Comment by AlexanderKurtz** [comment]
I don't think it is possible to replace ```laik_tcp_debug``` with ```laik_log()``` because it misses significant features:

 * does not work before an instance is created (this is critical for me!)
 * does not automatically gather PID, function, and line which are invaluable for debugging
 * can not be removed at compile time entirely
 * does not print its output tab-separated, so it can be post-processed if required

As for the error handling, I didn't see any such feature in LAIK core, did I miss anything?

**Comment by weidendo** [comment]
We always can extend/change laik_log stuff as needed.

> does not work before an instance is created (this is critical for me!)

I think the instance is only needed to be able to do filtering according to rank number, and generate the logging prefix string (which includes phase names and iteration counter). If this is not (yet) available, we can just replace it with "-1" as rank and iteration counter, and an empty string for the phase name (which is the default anyway).

> does not automatically gather PID, function, and line which are invaluable for debugging

Should be easy to add on top?
Also, PID/hostname/function/line may be useful in the prefix string anyway for all laik_log output.
We can extend the LAIK_LOG syntax to allow to specify what should go into the prefix.
And I think you also want to specify component name ("tcp"), and be able to filter output according
to components.

> can not be removed at compile time entirely

Just a matter of using as macro...

> does not print its output tab-separated, so it can be post-processed if required

What did you use it for?
I always can grep/search in the text output, which is essential with 10MB log output.
And the prefix allows for sensible sorting of phases with interleaved output per rank.

> As for the error handling, I didn't see any such feature in LAIK core, did I miss anything?

Currently LAIK just panics and exits on error. What do you need on top of that?
At some point, LAIK functions should be able to return errors to allow the application
to handle them. Is it about such functionality?

**Comment by AlexanderKurtz** [comment]
Regarding the debugging output: Your argument of just ```grep``` is precisely my problem: Debugging the TCP backend is a multi-threaded **and** multi-process problem which produces **huge** amounts of debug output. Quite frankly, ```grep``` just doesn't cut it here. I often times found myself importing the debug output into LibreOffice, setting up some conditional formatting and/or sorting so I could get a grasp of what's going on. Unless ```laik_log()``` can do this (and work properly without an instance), I just **can't** use it.

Regarding making ```laik_log()``` a macro: I have not tested this, but I strongly suspect that several variables would then become unused, which would make compilation fail. Also, keep in mind that "disabling" by macro means also not evaluating any contained expressions, which means those expressions must not have any side effects (see also #92). 

Regarding the error handling: No, just ```abort()```ing on error is not error handling. Not even in the development phase. The main problem I have with this, is that throws away all the information from the call stack: If a function ```foo``` encounters an error, it's rarely ```foo``` who is at fault, but most likely something further up the call chain. Which is why my error handling has an error stack, where every function can add its own context before passing the error further up. Oh, and how should ```foo``` know whether the error it encountered is fatal, so it can abort, or whether the caller can deal with the error? E.g. my rank determination logic works by trying to bind all the possible addresses and seeing which succeeded...

**TL;DR:** LAIK needs structured (i.e. machine-readable) logging and a proper (and consistent) error reporting system. It simply has grown too complex for ```fprintf()``` and ```abort()``` to be enough. Honestly, I fully expect that we'll eventually have to bite the bullet and switch to C++ (or Rust?) and Boost (or something similar), because we **need** the features they provide.

**Comment by weidendo** [comment]
This is not a question of "if", but "how". It makes absolutely no sense to use different log/error stuff within one library. The current state is bad. There is no hurry, but the goal must be clear.

Now on to the details...

> Debugging the TCP backend is a multi-threaded and multi-process...

Same for the MPI backend. It only will get more tricky with multiple backends used at the same time.
To understand complex bugs, my solution was sorting with the help of prefix strings, to get related information onto one screen. I understand that you imported debug output as CSV file into LibreOffice? What did it help for?

> macro

Same as with assert. We will get it working!

> abort ... call stack

This comment was about panic-like errors. You want to use a debugger for that, and then,
abort is exactly the right thing. Or generate a core dump for analysis in the debugger.
And of course, correct error handling is important and good to have throughout LAIK.

> switch to C++ (or Rust?)

Requirement: easy to compile out-of-the-box on as much as possible HPC systems with possibly old and conservative build stacks.
